Examples/JavaScript/JavaScript Essentials/Code Flow.ts

```diff
@@ -0,0 +1,60 @@
+//// { order: 3, compiler: { strictNullChecks: true } }
+
+// How code flows inside your JavaScript file can affect what the types for your
+// JavaScript looks like. If statements are great place to start here, because 
+// they provide a way to confirm a particular shape of an object.
+
+const users = [{ name: "Ahmed" }, { name: "Gemma" }, { name: "Jon" }]
+
+// We're going to look to see if we can find a user
+const jon = users.find(u => u.name === "jon")
+
+// If you hover over the three different jon's below, you'll see how the
+// type definitions change depending on where the word is located:
+
+if (jon) {
+  jon
+} else {
+  jon
+}
+
+// This uses a TypeScript feature called union types. A union type is a
+// way to declare that an object could be one of many things. JavaScript's
+// dynamic nature means a lot of function receive and return objects of
+// different types and you need to be able to figure out which one you have.
+
+// In the above case find could fail, and in that case you don't have an
+// object. This is written like: { name:string } | undefined 
+// The pipe acts as the separator between different types.
+
+// You can use this in a few ways, lets start by looking at an array
+// where the variables are not all the same.
+
+const identifiers = ["Hello", "World", 24, 19]
+
+// Could be a vinyl or cassette
```

Comment:
comment seems leftover from a cooler example

Reply:
Yeah, it turned out it was really hard to have an example like:

```ts
const products = [{ name: "Cassette", rewound: true }, { name: "Vinyl", age: 32 }]

// Could be a vinyl or cassette
const firstProduct = [products][0]

// We can use the JavaScript "x in y" syntax to check for one
// property which is only in vinyl. You can hover to see how
// firstProduct changes between different locations
if ("rewound" in firstProduct) {
  firstProduct
} else {
  firstProduct
}
```

because of how we union the objects in the array: 

```ts
const products: ({
    name: string;
    rewound: boolean;
    age?: undefined;
} | {
    name: string;
    age: number;
    rewound?: undefined;
})[]
```

They need to have different type roots for this example to work I think. Maybe I can make that pretty cheap conceptually by using the `typeof` above to make the type definition next to the instance.